lib/livebook/session.ex

```diff
@@ -796,16 +838,53 @@ defmodule Livebook.Session do
   end
 
   defp session_tmp_dir(session_id) do
-    tmp_dir = System.tmp_dir!() |> Path.expand()
-    path = Path.join([tmp_dir, "livebook", "sessions", session_id]) <> "/"
-    FileSystem.File.local(path)
+    livebook_tmp_path()
+    |> Path.join("sessions/#{session_id}")
+    |> FileSystem.Utils.ensure_dir_path()
+    |> FileSystem.File.local()
   end
 
   defp cleanup_tmp_dir(session_id) do
     tmp_dir = session_tmp_dir(session_id)
     FileSystem.File.remove(tmp_dir)
   end
 
+  defp local_assets_path(hash) do
+    Path.join([livebook_tmp_path(), "assets", encode_path_component(hash)])
+  end
+
+  @doc """
+  Returns a local path to asset matching the given
+  hash and path.
+
+  The file is not guaranteed to exist. See `fetch_assets/2`
+  for fetching assets through a particular session.
+
+  The path is expected to be a simple relative path
+  within the assets directory, otherwise an error is
+  returned.
+  """
+  @spec local_asset_path(String.t(), String.t()) :: {:ok, String.t()} | :error
+  def local_asset_path(hash, asset_path) do
+    assets_path = local_assets_path(hash)
+    local_asset_path = Path.expand(asset_path, assets_path)
+
+    if String.starts_with?(local_asset_path, assets_path <> "/") do
+      {:ok, local_asset_path}
+    else
+      :error
+    end
+  end
+
+  defp encode_path_component(component) do
+    String.replace(component, [".", "/"], "_")
```

Comment:
From where `.` and `/` are coming from? If it is a hash, we shouldn't be seeing a `.` in it?

Reply:
Or is this for security to not allow file system traversals? If so, also add `:` for Windows.

Reply:
Oh, we should add both `\\` and `:` for Windows.

Reply:
Yeah, `hash` comes from the runtime, we know that Kino specifically sends the correct hash, but it may not be the case.